### ISSUE TITLE
Implement ALLOCATE AVAILABLE

### ISSUE THREAD

**Issue by ToddFincannon** [climateinteractive/SDEverywhere#100]
Implement the [ALLOCATE AVAILABLE](https://www.vensim.com/documentation/fn_allocate_available.html) Vensim function limited to the case we currently need for a model: a one-to-many allocation with a normal curve type.

**Comment by ToddFincannon** [comment]
Jeff Rissman (@jrissman) provided an overview of how the [ALLOCATE AVAILABLE](https://www.vensim.com/documentation/fn_allocate_available.html) Vensim function works based on his experience using it. This formed the basis for the implmentation described below.

The function allocates a scalar quantity of an available resource to an array of requesters. It is calculated independently at each time step. Whether the agents are demanders or suppliers doesn't matter for the purposes of the calculation. We focus exclusively here on one priority profile type, the normal curve, that is most often used.

### Inputs

All arrays are subscripted by a dimension with one index for each requester.

_requested quantity_ array
_priority profile_ array
_available resource_

where the priority profile has an additional subscript with the following indices:

_ptype_ = 3 (the normal curve type)
_ppriority_ = mean of the normal curve
_pwidth_ = standard deviation of the normal curve
_pextra_ (not used)

Each request is represented by a normal curve having mean (μ) and standard deviation (σ) parameters. The requested quantity is the area under the curve. The priority is indicated by μ, which locates the maximum of the curve on the x axis. We integrate the area under the curve from x to +∞ so that requests with higher priorities (and thus greater means) get larger allocations. The standard deviation σ gives the spread of the curve, with smaller values more locally concentrated around the mean.

### Math functions

The mathematical underpinning of the calculation is the well-known formula for the normal curve with μ and σ parameters. Because we are scaling by the area under the curve, we also need an approximation to the integral, for which there is no closed-form solution. A simple and accurate formula is the approximation due to Zelen & Severo (1964) in the [Handbook Of Mathematical Functions](https://secure.math.ubc.ca/~cbm/aands/page_932.htm), Abramowitz and Stegun, section 26.2.17.

There is another set of approximations in the GNU Scientific Library (GSL) that customizes the approach based on which part of the normal curve is being integrated. The additional level of accuracy was judged to be unnecessary for this implementation. The source code was still useful for understanding how to apply the μ and σ parameters in the integration.

Because the integral of the unit normal curve with σ=1 at +∞ is equal to one, we simply multiply by the requested quantity constant to get that area under the curve.

### The algorithm

Having these prerequisites in hand, we are ready to implement ALLOCATE AVAILABLE.

```
For each time step:

    Find the available resource at the current time. Limit it to the total quantity from all requesters.

    Consider x values from the lowest priority minus 5 to the highest priority plus 5. Start with the x value in the middle.

        For each requester:

            Calculate the area under the curve from +∞ down to x using the normal CDF with the given μ and σ, scaled by the requested allocation.

            Save the area as the allocation for the requester.

        If the total allocations are equal to the available resource up to an epsilon precsion, stop and return the allocations.

        Otherwise, increase or decrease x depending on whether total allocations are over or under the available resource.
```

The scan limits are set to ±5 around the minimum and maximum mean values from the observation that this accounts for almost all the area under the normal curves. The epsilon is set to 1e-6 to get a highly precise result. This is found with relatively few evaluations by halving the x step size on each loop, that is, by performing a binary search.

In some cases, halving the x step size prevents x from reaching the target value where the total allocations equal the total available resource. What we want is for x to bounce around both sides of the target until it converges on it. But the binary search may approach from one side and rapidly decrease the x step size, falling short of the target. To prevent this, we detect if two successive steps have gone in the same direction, and stop halving the x step size until we cross over the target.

**Comment by ToddFincannon** [comment]
The `ALLOCATE AVAILABLE` function is complete with the following caveats.

### The maximum number of requesters is ~60~ 80

The C code for the `_ALLOCATE_AVAILABLE` function was simplified by using a static buffer for allocation results. This is arbitrarily limited to ~60~ 80 requesters, which is enough for current purposes.

(Updated by Chris on 2026.8.7: The max was changed from 60 to 80 in #837.)

### Vensim result comparisons must use reduced precision

SDEverywhere calculates the `ALLOCATE AVAILABLE` function with greater precision than Vensim. For instance, at time=0.5 in the test model, the "total supply available" is 0.416667. Our allocations add up to that value with 1e-6 precision, but Vensim's allocations add up to 0.416658, which is quite a bit off, but good enough for System Dynamics models.

I investigated further by using the inverse cdf function in the GSL library to find the x value for each allocation. Ours is exactly the same for all three (by design), but Vensim's are all different: 4.28978, 4.29089, 4.29132.

What this means is that we will not be able to reproduce Vensim results for `ALLOCATE AVAILABLE` with as much precision as we would like. We need to relax the test for the allocate model from 1e-4 to 1e-2 precision.

### ~The only supported ptype value is 3 (normal)~

See the [Specifying the Curves](https://www.vensim.com/documentation/24335.html) page in the Vensim documentation.

~SDEverywhere currently only implements the `ptype=3` case.  We will report a warning or error message if other values are used.
(Added by Chris on 2024.11.28.)~

(Updated by Chris on 2026.8.7: Support for other curve types has been added as part of #839.)

### ~The priority profile argument subscripts must be in alphanumeric order~

~The second argument of the `ALLOCATE AVAILABLE` function indicates the priority profile array by giving the name of the first subscript element. This is necessary in Vensim because naming the array itself on the RHS would require the array to also appear on the LHS. However, we don't want to assign to this array. This is basically a hack in the Vensim function signature.~

~This causes a problem for SDEverywhere, because the priority profile argument has two subscripts, and SDEverywhere converts them to a canonical form. That is, the subscript families are determined, converted to C identifiers, and then sorted in alphanumeric order. This is necessary because Vensim subscripts are not required to occur in the same order in every reference in an equation, but we need them to be in the same order in our C code. The problem is that the priority profile argument hack means we can't reorder the subscripts. The second subscript must be the pprofile subscript for the hack to work.~

~There is no clean solution to this problem in SDEverywhere. Instead, we need to place a restriction on the usage of `ALLOCATE AVAILABLE`, where the subscript families in the second argument are in alphanumeric order. This is reflected in the `allocate` test model.~

(Updated by Chris on 2024.11.28: This caveat no longer applies now that alphabetical normalization of subscripts is being removed as part of the fixes for #179 and #377.)
